design/history/exploration-reports/2020.10-data-model.md

```diff
@@ -0,0 +1,91 @@
+
+Going from programming language to Data Model (and back) is programming language specific. It depends on how a programming language works and which types are implemented. Going from Codec to Date Model is Codec specific. Some Codecs might implement only a subset of the Data Model Kinds or perhaps even a superset.
+
+On the Data Model layer we don't talk about memory layouts or how things should be serialized. We only talk about the properties of the Kinds. How they are implemented is then up to the library. The memory layout and type depends on the programming language, the serialization is the concern of IPLD Codecs. We try to minimize the number of Kinds to the minimum, while still having a useful system, where every Kind has its distinct properties.
+
+
+Primitive Kinds
+---------------
+
+### Null
+
+The `null` Kind has a cardinality of 1.
+
+
+### Boolean
+
+A boolean is either `true` or `false`. It has a cardinality of 2.
+
+
+### Integer
+
+A numeric value without fractions. It can be positive or negative.
+
+
+### Float
+
+A numeric value that represents a fraction.
+
+
+### Link
+
+A link is an identifier to some content-addressed data. It has the same properties as a CID.
+
+
+Compound Kinds
+--------------
+
+### Byte-Array
+
+A fixed sized sequence of 8-bit bytes. We don't define a primitive `Byte` Kind, that would be redundant as it can be represented as a single element Byte-Array.
+
+
+### Text
+
+Text is a fixed size sequence of characters. Characters are different from bytes as they have a specific encoding, so that the Text can be interpreted in a meaningful way. For Text without a specified encoding, or if you don't care about the encoding, use `Byte-Arrays` instead. To keep things simple, there is no support for arbitrary encodings, the only valid encoding Unicode.
+
+Just as with `Byte-Array`s, there is no `Character` Kind specified, use `Text` with a single character instead.
+
+
+Container Kinds
+---------------
+
+### Lists
+
+A fixed sized sequence of arbitrary Primitive or Container Kinds.
+
+
+### Map
+
+A collection of key value pairs. The key is of Kind `Byte-Array`, the values can be of any Kind.
+
+---
+
+Appendix
+--------
+
+**Q: Why is the `Text` Kind Unicode and not UTF-8?**
+
+**A:** The data model doesn't talk about memory layouts. Your programming language might encode Unicode as UTF-16 (e.g. Java, JavaScript) and that's totally fine. Your Codec then defines how to serialize things, which is then likely UTF-8, but any other encoding would also be possible
+
+
+**Q: My programming language only supports strings as map keys**
+
+**A:** Depending on your language that might not be a problem. The Data Model talks about Kinds in abstract. This means that a Kind does not necessarily be the same programming language type in all cases.
```

Comment:
The byte-keys thing seems to be the most controversial component in this PR, I assume that's the primary intent since it's the largest diversion from what we have now.

The logic used here, and the example included below, could be flipped the other way to support our current standard: maps have string keys except in the languages that have trouble with the loose-unicode view, in which case it's welcome to interpret them as bytes and use them that way as long as users are made aware of the differences.

Reply:
> The byte-keys thing seems to be the most controversial component in this PR

I had hope that wouldn't be the case. I had hoped that the most controversial part is that I talk about the Data Model in an abstract way, which is independent of the programming language or the Codec. So far we often convoluted those concerns and I hope we can make clearer distinctions.